src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -238,28 +243,190 @@ private SNITCPHandle CreateTcpHandle(string fullServerName, long timerExpire, ob
+        {
+            responsePacket = null;
+
+            if (browserEP == null || String.IsNullOrWhiteSpace(instanceName))
+            {
+                return false;
+            }
+
+            List<byte> byteList = new List<byte>(new byte[] { 0x04 });
+            byteList.AddRange(Encoding.ASCII.GetBytes(instanceName));
+            byte[] requestPacket = byteList.ToArray();
+
+            bool success = true;
+            UdpClient client = null;
+            try
+            {
+                client = new UdpClient(AddressFamily.InterNetwork);
+                Task<int> sendTask = client.SendAsync(requestPacket, requestPacket.Length, browserEP);
+                Task<UdpReceiveResult> receiveTask = null;
+                if (sendTask.Wait(1000) && (receiveTask = client.ReceiveAsync()).Wait(1000))
+                {
+
+                    responsePacket = receiveTask.Result.Buffer;
+                }
+                else
+                {
+                    success = false;
+                }
+            }
+            catch
+            {
+                success = false;
+            }
+            finally
+            {
+                if (client != null)
+                {
+                    client.Dispose();
+                }
+            }
+
+            return success;
+        }
+
+        /// <summary>
+        /// Finds port number for given instance name.
+        /// </summary>
+        /// <param name="browserHostname">SQL Sever Browser hostname</param>
+        /// <param name="instanceName">instance name to find port number</param>
+        /// <param name="instancePort">port number for given instance name</param>
+        /// <param name="browserPort">>SQL Sever Browser port. Default value is 1434</param>
+        /// <returns>bool value showing if operation is successful</returns>
+        private static bool GetPortByInstanceName(string browserHostname, string instanceName, out int instancePort, int browserPort = 1434)
+        {
+            instancePort = -1;
+
+            if (String.IsNullOrWhiteSpace(browserHostname) || String.IsNullOrWhiteSpace(instanceName))
+            {
+                return false;
+            }
+
+            IPAddress serverIPAddress = null;
+            if (!GetIPAdress(browserHostname, out serverIPAddress))
+            {
+                return false;
+            }
+
+            IPEndPoint browserEP = new IPEndPoint(serverIPAddress, browserPort);
+
+            byte[] responsePacket = null;
+            if (!Send_CLNT_UCAST_INST(browserEP, instanceName, out responsePacket) || responsePacket == null || responsePacket.Length <= 3)
+            {
+                return false;
+            }
+
+            string serverMessage = Encoding.ASCII.GetString(responsePacket, 3, responsePacket.Length - 3);
+            string[] elements = serverMessage.Split(';');
+            int tcpIndex = (new List<string>(elements)).IndexOf("tcp");
```

Comment:
Nit: No need to allocate and copy `elements` into a new `List<string>`. Just use `Array.IndexOf`, e.g.:

```c#
int tcpIndex = Array.IndexOf(elements, "tcp");
```